File: python/algos/single_number.py
from typing import Optional


def single_number(numbers: list[int]) -> int:
    occurrences = {}
    for value in numbers:
        occurrences[value] = 1 + occurrences.get(value, 0)
    for value in numbers:
        if occurrences[value] == 1:
            return value


if __name__ == '__main__':
    assert single_number([2, 2, 1]) == 1
    assert single_number([4, 1, 2, 1, 2]) == 4
    assert single_number([1]) == 1

File: rust/src/algos/single_number.rs
use std::collections::HashMap;

pub fn single_number(numbers: Vec<i32>) -> i32 {
    let mut occurrences: HashMap<i32, i32> = HashMap::new();
    for number in numbers.iter() {
        *occurrences.entry(*number).or_insert(0) += 1;
    }
    for (key, number) in occurrences.iter() {
        if *number == 1 {
            return *key
        }
    }
    0
}